cli/daemon/run/proc_groups.go

```diff
@@ -0,0 +1,297 @@
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to generate runtime config")
+	}
+	runtimeJSON, _ := json.Marshal(runtimeCfg)
+
+	cmd := exec.CommandContext(pg.ctx, binPath)
+	envs := append([]string{
+		"ENCORE_RUNTIME_CONFIG=" + base64.RawURLEncoding.EncodeToString(runtimeJSON),
+		"ENCORE_APP_SECRETS=" + encodeSecretsEnv(secrets),
+	},
+		environ...,
+	)
+	for serviceName, cfgString := range configs {
+		envs = append(envs, "ENCORE_CFG_"+strings.ToUpper(serviceName)+"="+base64.RawURLEncoding.EncodeToString([]byte(cfgString)))
+	}
+
+	cmd.Env = envs
+	cmd.Dir = filepath.Join(pg.Run.App.Root(), pg.workingDir)
+
+	// Proxy stdout and stderr to the given app logger, if any.
+	if l := pg.logger; l != nil {
+		cmd.Stdout = newLogWriter(pg.Run, l.RunStdout)
+		cmd.Stderr = newLogWriter(pg.Run, l.RunStderr)
+	}
+
+	p.cmd = cmd
+
+	// Set up extra file descriptors for communicating requests/responses:
+	// - reqRd is for reading incoming requests (handed over procchild)
+	// - reqWr is for writing incoming requests
+	// - respRd is for reading responses
+	// - respWr is for writing responses (handed over to proc)
+	reqRd, reqWr, err1 := os.Pipe()
+	respRd, respWr, err2 := os.Pipe()
+	defer func() {
+		// Close all the files if we return an error.
+		if err != nil {
+			closeAll(reqRd, reqWr, respRd, respWr)
+		}
+	}()
+	if err := firstErr(err1, err2); err != nil {
+		return errors.Wrap(err, "could not create request & response pipes")
+	} else if err := xos.ArrangeExtraFiles(cmd, reqRd, respWr); err != nil {
+		return errors.Wrap(err, "could not handle over request & response pipes")
+	}
+	p.reqWr = reqWr
+	p.respRd = respRd
+	p.closeOnStart = []io.Closer{reqRd, respWr}
+
+	rwc := &struct {
+		io.ReadCloser
+		io.Writer
+	}{
+		ReadCloser: io.NopCloser(respRd),
+		Writer:     reqWr,
+	}
+	p.Client, err = yamux.Client(rwc, yamux.DefaultConfig())
+	if err != nil {
+		return errors.Wrap(err, "could not initialize connection")
+	}
+
+	return nil
+}
+
+// Proc represents a single Encore process running within a [ProcGroup].
+type Proc struct {
+	group *ProcGroup     // The group this process belongs to
+	log   zerolog.Logger // The logger for this process
+	exit  chan struct{}  // closed when the process has exited
+	cmd   *exec.Cmd      // The command for this specific process
+
+	Started   atomic.Bool    // whether the process has started
+	StartedAt time.Time      // when the process started
+	Pid       int            // the OS process id
+	Client    *yamux.Session // the client connection to the process
+
+	reqWr        *os.File
+	respRd       *os.File
+	closeOnStart []io.Closer
+}
+
+// Start starts the process and returns immediately.
+//
+// If the process has already been started, this is a no-op.
+func (p *Proc) Start() error {
+	if !p.Started.CompareAndSwap(false, true) {
+		return nil
+	}
+
+	p.group.procMu.Lock()
+	defer p.group.procMu.Unlock()
+
+	if err := p.cmd.Start(); err != nil {
```

Comment:
We can lock the proc group after `p.cmd.Start()` I guess?

Reply:
I did it before the `Start` so `runningProcs` would be correct for all readers at the time of read, if we took the lock after `Start` it might be possible for a reader to read `0` just before we increment it to `1`.